### PR TITLE
Some shortcuts

### PR DESCRIPTION
**Pull request by johnno1962**
Some more and less zany ideas to cut down on code for common usage patterns. Very optional!

### CODE REVIEW

File: SwiftLSPClient/Types/TextDocumentIndentifier.swift [hunk 1]
```diff
@@ -14,6 +14,10 @@ public struct TextDocumentIdentifier {
     public init(uri: DocumentUri) {
         self.uri = uri
     }
+
+    public init(path: String) {
+        self.uri = URL(fileURLWithPath: path).absoluteString
+    }
```
Comment (mattmassicotte) [inline SwiftLSPClient/Types/TextDocumentIndentifier.swift]: 👍 definitely looks useful

File: SwiftLSPClient/Types/Features/Reference.swift [hunk 2]
```diff
@@ -21,7 +21,8 @@ public struct ReferenceParams: Codable {
     public let position: Position
     public let context: ReferenceContext
 
-    public init(textDocument: TextDocumentIdentifier, position: Position, context: ReferenceContext) {
+    public init(textDocument: TextDocumentIdentifier, position: Position,
+                context: ReferenceContext = ReferenceContext(includeDeclaration: false)) {
```
Comment (mattmassicotte) [inline SwiftLSPClient/Types/Features/Reference.swift]: What do you think about just adding a ` = false` to the next init method? Much shorter, and semantically equivalent as long as `ReferenceContext` doesn't change.

File: SwiftLSPClient/Types/Basic.swift [hunk 3]
```diff
@@ -58,7 +58,7 @@ public enum LanguageIdentifier: String, Codable, CaseIterable {
     case objc = "objective-c"
     case objcpp = "objective-cpp"
 
-    static let extensions = [
+    static let fileExtensions = [
```
Comment (mattmassicotte) [inline SwiftLSPClient/Types/Basic.swift]: Thanks for doing this!

File: SwiftLSPClient/Types/Basic.swift [hunk 4]
```diff
@@ -83,6 +83,20 @@ public struct TextDocumentItem: Codable {
...
+        let ext = url.pathExtension
+        guard let languageID = LanguageIdentifier
+            .fileExtensions[ext] else {
+                throw NSError(domain: "SwiftLSPClient", code: -1,
+                              userInfo: ["Invalid source type \(ext)": path])
+        }
+        self.languageId = languageID
+        self.version = 1
+        self.text = try String(contentsOf: url)
+    }
```
Comment (mattmassicotte) [inline SwiftLSPClient/Types/Basic.swift]: There's a lot happening in this method. How about something like this:

- add an init similar to what do you with `TextDocumentIdentifier`
- add a static method to `LanguageIdentifier` to compute a languageID given a url
- finally, wrap this all up in something like `init(contentsAt: path String, version: Int = 1)`
- Also might be nicer to wrap up that error into an Error type. A little more swifty, and makes it easier to surface the contextual information (that the extension isn't in that dictionary)

This should achieve exactly what you are after, keep some consistency with the change to `TextDocumentIdentifier` and also add some flexibility for other uses.

File: SwiftLSPClient/Types/Features/Reference.swift (hunk 2 above)
Comment (johnno1962) [reply SwiftLSPClient/Types/Features/Reference.swift]: 👍 Done.

File: SwiftLSPClient/Types/Basic.swift (hunk 3 above)
Comment (johnno1962) [reply SwiftLSPClient/Types/Basic.swift]: No problem!

File: SwiftLSPClient/Types/Basic.swift (hunk 4 above)
Comment (johnno1962) [reply SwiftLSPClient/Types/Basic.swift]: I don't know if you are still awake but I've pushed another commit looking to address what you are after and the code seems to be the better for it.

Comment (johnno1962) [reply SwiftLSPClient/Types/Basic.swift]: I just saw another thread crash (1 in a day) so my locks aren't quite up to scratch.

File: SwiftLSPClient/Types/Basic.swift [hunk 5]
```diff
@@ -83,6 +96,14 @@ public struct TextDocumentItem: Codable {
         self.version = version
         self.text = text
     }
+
+    public init(contentsAt path: String, version: Int = 1) throws {
```
Comment (mattmassicotte) [inline SwiftLSPClient/Types/Basic.swift]: I've just noticed that String uses the init form `String(contentsOfFile: path)`. Probably best to match that convention, and rename this parameter `contentsOfFile`.

File: SwiftLSPClient/Types/Features/Reference.swift [hunk 6]
```diff
@@ -21,7 +21,8 @@ public struct ReferenceParams: Codable {
     public let position: Position
     public let context: ReferenceContext
 
-    public init(textDocument: TextDocumentIdentifier, position: Position, context: ReferenceContext) {
+    public init(textDocument: TextDocumentIdentifier, position: Position,
+                context: ReferenceContext = ReferenceContext()) {
```
Comment (mattmassicotte) [inline SwiftLSPClient/Types/Features/Reference.swift]: What you've got here will work just fine. However, if instead you modify the following init method:

    public init(textDocument: TextDocumentIdentifier, position: Position, includeDeclaration: Bool)

to:

    public init(textDocument: TextDocumentIdentifier, position: Position, includeDeclaration: Bool = false)

You get the same API without affecting how `ReferenceContext` works. What do you think?

Comment (johnno1962) [reply SwiftLSPClient/Types/Features/Reference.swift]: D'oh, misunderstood you there. Committed a new commit.

File: SwiftLSPClient/Types/Basic.swift [hunk 7]
```diff
@@ -67,8 +67,20 @@ public enum LanguageIdentifier: String, Codable, CaseIterable {
         "cpp": .cpp,
         "m": .objc,
         "mm": .objcpp,
-        "h": .objcpp
     ]
+
+    public struct UnsupportedFileType: Error {
+        let msg: String
+    }
```
Comment (mattmassicotte) [inline SwiftLSPClient/Types/Basic.swift]: I was thinking about this more. There are actually a bunch of other kinds of parameter validation that could be done. Perhaps instead we can do something like this:

    public enum LanguageServerParameterError: Error {
        case unsupportedFileExtension(String)
    }

This allows for much better communication of the semantics in the error, while also making it possible to expose all kinds of other parameter errors in the future without custom types for each one. This is a pretty common idiom for swift throwing errors, and it's already in use in the library with `LanguageServerError`. You into that?

Comment (johnno1962) [reply SwiftLSPClient/Types/Basic.swift]: Done 👍

**Comment by johnno1962** [comment]
Bother, I merged the locks changes onto the branch for this PR and now they are included. Do you want me to clean this up? The server is reliable with it though don't know if it will cause deadlocks. It only affects one file: `SwiftLSPClient/JSONRPC/ProtocolTransport.swift`

**Comment by johnno1962** [comment]
I've reversed the changes to SwiftLSPClient/JSONRPC/ProtocolTransport.swift

**Comment by johnno1962** [comment]
Thought on this: https://github.com/johnno1962/SwiftLSPClient/commit/f7a4cdfca5d82eeff1e8187206725d4671341338? I’m experimenting with a generic message api keyed on the ...Params struct. It requires separate types for each request but you’d then be able to offer sync/async interfaces with two entry points.

**Comment by mattmassicotte** [comment]
This is looking great now! Just have to remove some of the locking stuff that snuck in :)

**Comment by johnno1962** [comment]
Ooops I though I'd removed them. Should be fine now. Apologies. Don't forget squash.

**Comment by mattmassicotte** [comment]
This looks great! Thanks for doing this. Really makes me realize how many other areas can be made more convenient.

**Comment by johnno1962** [comment]
Thanks for merging this. It’s nice to be synced up. I see you added comments about the issue I’m seeing which is going to be very helpful. Great project !

**Comment by mattmassicotte** [comment]
@johnno1962 absolutely! This library is a critical aspect of my application, so it's important to me :)

**Pull request merged**